### PR TITLE
feat: enable right filtering when specifing hostRules with github-api using datasources

### PR DESCRIPTION
**Pull request by jeskosda**
## Changes:

As discussed in: https://github.com/renovatebot/renovate/issues/10737

1. Pass the datasources to GithubHttp to enable right filtering when specifing hostRules with github-api using datasources like github-releases etc. 
2. Adding generic hostRule for github platform to match all github-api request from datasource per default to enable backwards compatiblity.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [X] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/http/github.ts [hunk 1]
```diff
@@ -152,8 +152,14 @@ function constructAcceptString(input?: any): string {
...
-  constructor(options?: GithubHttpOptions) {
-    super(PLATFORM_TYPE_GITHUB, options);
+  constructor({
+    hostType = PLATFORM_TYPE_GITHUB,
+    options,
+  }: {
+    hostType?: string;
+    options?: GithubHttpOptions;
+  }) {
+    super(hostType, options);
```
Comment (viceice) [inline lib/util/http/github.ts:162]: Use optional second parameter instead, so most code can fallback to exiting working `hostType`

File: lib/platform/github/index.ts [hunk 2]
```diff
@@ -396,24 +403,24 @@ export async function initRepo({
...
   parsedEndpoint.host = parsedEndpoint.host.replace(
     'api.github.com',
     'github.com'
   );
   parsedEndpoint.pathname = config.repository + '.git';
-  const url = URL.format(parsedEndpoint);
+  const url = parsedEndpoint.toString();
   await git.initRepo({
     ...config,
     url,
```
Comment (viceice) [inline lib/platform/github/index.ts:426]: Please revert unrelated changes

File: lib/util/http/github.ts [hunk 3]
```diff
@@ -152,8 +152,11 @@ function constructAcceptString(input?: any): string {
...
 
 export class GithubHttp extends Http<GithubHttpOptions, GithubHttpOptions> {
-  constructor(options?: GithubHttpOptions) {
-    super(PLATFORM_TYPE_GITHUB, options);
+  constructor(
+    options?: GithubHttpOptions,
+    hostType: string = PLATFORM_TYPE_GITHUB
+  ) {
+    super(hostType, options);
   }
```
Comment (viceice) [inline lib/util/http/github.ts:160]: Can we flip the order? We never passed any `options` yet, so `hostType` is more mandatory and should com first.

We can refactor this in a later pr for the other HTTP classes

File: lib/util/http/auth.ts [hunk 4]
```diff
@@ -7,6 +7,14 @@ import {
 } from '../../constants/platforms';
 import { GotOptions } from './types';
 
+// Using the datasource constants would result into a circular dependency
+const GITHUB_HOST_TYPES = [
+  PLATFORM_TYPE_GITHUB,
+  'github-releases',
+  'github-tags',
+  'pod',
```
Comment (viceice) [inline lib/util/http/auth.ts:15]: I don't like this change, makes it pretty complicated to get all those types in sync.

Please move this to `lib/types/github.ts`. We should make some typescript assertions, so that this works if anything is changes. @JamieMagee Do you have an idea how to get this properly typed?

File: lib/util/http/auth.spec.ts [hunk 5]
```diff
@@ -69,6 +70,45 @@ describe(getName(), () => {
...
+      expect(opts).toMatchInlineSnapshot(`
+        Object {
+          "headers": Object {
+            "authorization": "token XXXX",
+          },
+          "hostType": "github",
+          "token": "XXXX",
+        }
+      `);
+    });
```
Comment (viceice) [inline lib/util/http/auth.spec.ts:91]: Please use `toEqual` matcher, so this didn't get silently updated by jest snapshot updates.

File: lib/datasource/github-releases/index.ts [hunk 6]
```diff
@@ -2,6 +2,7 @@ import { logger } from '../../logger';
 import * as packageCache from '../../util/cache/package';
 import type { DigestConfig, GetReleasesConfig, ReleaseResult } from '../types';
 import {
+  id as GITHUB_RELEASES_ID,
```
Comment (viceice) [inline lib/datasource/github-releases/index.ts:5]:
```suggestion
  id,
```

File: lib/datasource/github-releases/index.ts [hunk 7]
```diff
@@ -11,7 +12,7 @@ import {
 import { findDigestAsset, mapDigestAssetToRelease } from './digest';
 import type { GithubRelease } from './types';
 
-export const id = 'github-releases';
+export const id = GITHUB_RELEASES_ID;
```
Comment (viceice) [inline lib/datasource/github-releases/index.ts:15]:
```suggestion
export { id };
```

Use simpler export.

File: lib/platform/github/index.spec.ts [hunk 8]
```diff
@@ -130,6 +130,16 @@ describe(getName(), () => {
       ).toMatchSnapshot();
       expect(httpMock.getTrace()).toMatchSnapshot();
     });
+    it('should add initialized platform with predefined generic host rule for github api', async () => {
```
Comment (viceice) [inline lib/platform/github/index.spec.ts:133]:
```suggestion

    it('should add initialized platform with predefined generic host rule for github api', async () => {
```

Readabillity, Add newline before each new test. Ignore existing

File: lib/platform/github/index.ts [hunk 9]
```diff
@@ -1,4 +1,4 @@
-import URL from 'url';
+import { URL } from 'url';
```
Comment (viceice) [inline lib/platform/github/index.ts:1]:
```suggestion
```

Use global `URL`, better use `lib/utils/url` helper

File: lib/util/http/auth.ts (hunk 4 above)
Comment (jeskosda) [reply lib/util/http/auth.ts:15]: Yes! Also don't like it! Could create `lib/types/github.ts` but what we need to ensure is that the datasource ids and the list of github host types don't diverge. 👍🏻

Comment (jeskosda) [reply lib/util/http/auth.ts:15]: Hey @JamieMagee, do you have any thoughts? Otherwise, I would do the suggested changes of @viceice

File: lib/platform/github/index.ts [hunk 10]
```diff
@@ -118,6 +118,13 @@ export async function initPlatform({
     gitAuthor: gitAuthor || discoveredGitAuthor,
     renovateUsername,
   };
+
+  // Generic github hostRule that per default all datasources using github api are enabled
+  const genericGithubHostRule = {
+    matchHost: parseUrl(defaults.endpoint).hostname,
+    token,
+  };
+  hostRules.add(genericGithubHostRule);
```
Comment (viceice) [inline lib/platform/github/index.ts:127]: @rarkins WDYT

File: lib/util/host-rules.spec.ts [hunk 11]
```diff
@@ -107,6 +110,68 @@ describe(getName(), () => {
           .token
       ).toBeUndefined();
     });
+    it('matches on specific path', () => {
```
Comment (viceice) [inline lib/util/host-rules.spec.ts:113]:
```suggestion

    it('matches on specific path', () => {
```

readabillity

File: lib/util/host-rules.spec.ts [hunk 12]
```diff
@@ -107,6 +110,68 @@ describe(getName(), () => {
...
+        token: 'overwrittenToken',
+      } as any);
+      expect(
+        find({
+          hostType: PLATFORM_TYPE_GITHUB,
+          url: 'https://api.github.com/repos/org-b/someRepo/tags?per_page=100',
+        }).token
+      ).toEqual('overwrittenToken');
+    });
+    it('matches for several hostTypes when no hostType rule is configured', () => {
```
Comment (viceice) [inline lib/util/host-rules.spec.ts:139]:
```suggestion

    it('matches for several hostTypes when no hostType rule is configured', () => {
```

File: lib/util/host-rules.spec.ts [hunk 13]
```diff
@@ -107,6 +110,68 @@ describe(getName(), () => {
...
+        }).token
+      ).toEqual('abc');
+      expect(
+        find({
+          hostType: 'github-releases',
+          url: 'https://api.github.com/repos/org-b/someRepo/tags?per_page=100',
+        }).token
+      ).toEqual('abc');
+    });
+    it('matches if hostType is configured and host rule is filtered with datasource', () => {
```
Comment (viceice) [inline lib/util/host-rules.spec.ts:157]:
```suggestion

    it('matches if hostType is configured and host rule is filtered with datasource', () => {
```

File: lib/util/host-rules.spec.ts [hunk 14]
```diff
@@ -107,6 +110,68 @@ describe(getName(), () => {
...
+        matchHost: 'https://api.github.com/repos/org-b/',
+        token: 'overwrittenToken',
+      } as any);
+      expect(
+        find({
+          hostType: 'github-tags',
+          url: 'https://api.github.com/repos/org-b/someRepo/tags?per_page=100',
+        }).token
+      ).toEqual('overwrittenToken');
+    });
```
Comment (viceice) [inline lib/util/host-rules.spec.ts:174]:
```suggestion
    });

```

File: lib/util/host-rules.spec.ts [hunk 15]
```diff
@@ -107,6 +110,68 @@ describe(getName(), () => {
...
+          hostType: 'github-releases',
+          url: 'https://api.github.com/repos/org-b/someRepo/tags?per_page=100',
+        }).token
+      ).toEqual('abc');
+    });
+    it('matches if hostType is configured and host rule is filtered with datasource', () => {
+      add({
+        hostType: PLATFORM_TYPE_GITHUB,
+        matchHost: 'https://api.github.com',
+        token: 'defaultToken',
```
Comment (viceice) [inline lib/util/host-rules.spec.ts:161]: Please use shorter token to not trigger codeql hardcoded credentials issue

see https://github.com/github/codeql/blob/221a25993854d41ea455e91e146fe5ea0e8bd0ff/javascript/ql/src/semmle/javascript/security/SensitiveActions.qll#L183

for sample token / passwords

File: lib/util/host-rules.spec.ts [hunk 16]
```diff
@@ -107,6 +110,68 @@ describe(getName(), () => {
...
+      });
+      add({
+        hostType: 'github-tags',
+        matchHost: 'https://api.github.com/repos/org-b/',
+        token: 'overwrittenToken',
+      } as any);
```
Comment (viceice) [inline lib/util/host-rules.spec.ts:167]:
```suggestion
      });
```

Cast shouldn't required

File: lib/util/http/github.spec.ts [hunk 17]
```diff
@@ -62,7 +63,22 @@ describe(getName(), () => {
       expect(req.headers.accept).toBe(
         'some-accept, application/vnd.github.machine-man-preview+json'
       );
+      expect(req.headers.authorization).toBe('token abc123');
     });
+    it('supports different datasources', async () => {
```
Comment (viceice) [inline lib/util/http/github.spec.ts:68]:
```suggestion

    it('supports different datasources', async () => {
```

File: lib/util/http/github.spec.ts [hunk 18]
```diff
@@ -62,7 +63,22 @@ describe(getName(), () => {
...
+      const githubApiDatasource = new GithubHttp({}, GITHUB_RELEASES_ID);
+      hostRules.add({ hostType: 'github', token: 'x-access-token:abc123' });
```
Comment (viceice) [inline lib/util/http/github.spec.ts:70]: see codeql comment above

File: lib/util/http/auth.ts (hunk 4 above)
Comment (JamieMagee) [reply lib/util/http/auth.ts:15]: I can't really think of a better way of doing this. It doesn't make sense, logically, for anything other than the datasource itself to own its own type/id.

File: lib/platform/github/index.ts (hunk 10 above)
Comment (rarkins) [reply lib/platform/github/index.ts:127]: I agree with this

**Review by viceice — DISMISSED** [review]

**Review by viceice — APPROVED** [review]
